### PR TITLE
conn, restore: paginate scan regions

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve? 

Paginate scan regions to avoid gRPC max message size limitation.

### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: pkg/restore/split_test.go [hunk 1]
```diff
@@ -142,16 +150,14 @@ func (c *testClient) GetOperator(ctx context.Context, regionID uint64) (*pdpb.Ge
...
-		}
-		if (len(region.Region.GetEndKey()) != 0 && bytes.Compare(region.Region.GetEndKey(), key) <= 0) ||
-			bytes.Compare(region.Region.GetStartKey(), endKey) > 0 {
-			continue
-		}
-		regions = append(regions, region)
+	infos := c.regionsInfo.ScanRange(key, endKey, limit)
+	regions := make([]*RegionInfo, 0, len(infos))
+	for _, info := range infos {
+		fmt.Printf("region %v\n", info)
```
Comment (kennytm) [inline pkg/restore/split_test.go]: Remove this line

File: pkg/restore/util.go [hunk 2]
```diff
@@ -333,3 +334,37 @@ func escapeTableName(cis model.CIStr) string {
...
+		log.Fatal("startKey >= endKey",
+			zap.Binary("startKey", startKey), zap.Binary("endKey", endKey))
+	}
+
+	for {
+		batch, err := client.ScanRegions(ctx, startKey, endKey, limit)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		fmt.Println("scan regions length", len(batch))
```
Comment (kennytm) [inline pkg/restore/util.go]: Ditto

File: pkg/restore/util.go [hunk 3]
```diff
@@ -324,3 +324,36 @@ func encodeKeyPrefix(key []byte) []byte {
...
+
+// paginateScanRegion scan regions with a limit pagination and
+// return all regions at once.
+// It reduces max gRPC message size.
+func paginateScanRegion(
+	ctx context.Context, client SplitClient, startKey, endKey []byte, limit int,
+) ([]*RegionInfo, error) {
+	regions := []*RegionInfo{}
+
+	if len(endKey) != 0 && bytes.Compare(startKey, endKey) >= 0 {
```
Comment (3pointer) [inline pkg/restore/util.go]: How about return err immediately and add test in TestPaginateScanRegion?

**Comment by kennytm** [comment]
```
[2020-02-24T07:25:55.574Z] [2020/02/24 15:25:55.320 +08:00] [ERROR] [import.go:213] ["ingest file failed"] [file="name:\"7_4_98_915f219e2da8023b412878692cbae735745fa2dc77068af29f236ddbfb394583_default.sst\" sha256:\"EH\\270\\177\\362\\0218\\250wZ}*\\236\\240\\367d\\221\\237\\266\\0205\\034\\353|u5w(\\222\\317m\\t\" start_key:\"t\\200\\000\\000\\000\\000\\000\\000k_r\\000\\000\\000\\000\\000\\000\\000\\000\" end_key:\"t\\200\\000\\000\\000\\000\\000\\000k_r\\377\\377\\377\\377\\377\\377\\377\\377\\000\" end_version:414849500052979718 crc64xor:1192681137856122040 total_kvs:10000 total_bytes:103510000 cf:\"default\" "] [range="start:\"t\\200\\000\\000\\000\\000\\000\\000\\377}_r\\200\\000\\000\\000\\000\\377\\000\\006\\205\\000\\000\\000\\000\\000\\372\" end:\"t\\200\\000\\000\\000\\000\\000\\000\\377}_r\\200\\000\\000\\000\\000\\377\\000\\006\\234\\000\\000\\000\\000\\000\\372\" "] [region="id:7529 start_key:\"t\\200\\000\\000\\000\\000\\000\\000\\377}_r\\200\\000\\000\\000\\000\\377\\000\\006\\205\\000\\000\\000\\000\\000\\372\" end_key:\"t\\200\\000\\000\\000\\000\\000\\000\\377}_r\\200\\000\\000\\000\\000\\377\\000\\006\\235\\000\\000\\000\\000\\000\\372\" region_epoch:<conf_ver:17 version:318 > peers:<id:7530 store_id:8 > peers:<id:7531 store_id:7 > peers:<id:7532 store_id:1 > "] [error="region not found"] [stack="github.com/pingcap/log.Error\n\t/go/pkg/mod/github.com/pingcap/log@v0.0.0-20200117041106-d28c14d3b1cd/global.go:42\ngithub.com/pingcap/br/pkg/restore.(*FileImporter).Import.func1\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/import.go:213\ngithub.com/pingcap/br/pkg/utils.WithRetry\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/utils/retry.go:27\ngithub.com/pingcap/br/pkg/restore.(*FileImporter).Import\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/import.go:163\ngithub.com/pingcap/br/pkg/restore.(*Client).RestoreFiles.func2\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/restore/client.go:296\ngithub.com/pingcap/br/pkg/utils.(*WorkerPool).Apply.func1\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/utils/worker.go:45"]
```

**Comment by kennytm** [comment]
```
[2020-03-03T05:52:30.486Z] [2020/03/03 13:52:29.963 +08:00] [WARN] [backoff.go:309] ["pdRPC backoffer.maxSleep 20000ms is exceeded, errors:\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:09.095769894+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:09.53067631+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:10.411808547+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:11.764856267+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:13.678756577+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:16.086870797+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:18.031000625+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:19.862126311+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:21.426172104+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:22.961369134+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:25.058575659+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:27.702821068+08:00\nregion not found for key \"mBootstra\\xffpKey\\x00\\x00\\x00\\x00\\xfb\\x00\\x00\\x00\\x00\\x00\\x00\\x00s\" at 2020-03-03T13:52:29.963828507+08:00"]
[2020-03-03T05:52:30.486Z] [2020/03/03 13:52:29.963 +08:00] [DEBUG] [txn.go:324] ["[kv] rollback txn"] [txnStartTS=415030241938374658]
[2020-03-03T05:52:30.487Z] [2020/03/03 13:52:29.963 +08:00] [FATAL] [session.go:1802] ["check bootstrapped failed"] [error="[tikv:9001]PD server timeout"] [errorVerbose="[tikv:9001]PD server timeout\ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20190809092503-95897b64e011/errors.go:174\ngithub.com/pingcap/errors.Trace\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20190809092503-95897b64e011/juju_adaptor.go:15\ngithub.com/pingcap/tidb/store/tikv.(*RegionCache).loadRegion\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/store/tikv/region_cache.go:773\ngithub.com/pingcap/tidb/store/tikv.(*RegionCache).findRegionByKey\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/store/tikv/region_cache.go:481\ngithub.com/pingcap/tidb/store/tikv.(*RegionCache).LocateKey\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/store/tikv/region_cache.go:441\ngithub.com/pingcap/tidb/store/tikv.(*tikvSnapshot).get\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/store/tikv/snapshot.go:321\ngithub.com/pingcap/tidb/store/tikv.(*tikvSnapshot).Get\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/store/tikv/snapshot.go:276\ngithub.com/pingcap/tidb/kv.(*unionStore).Get\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/kv/union_store.go:201\ngithub.com/pingcap/tidb/store/tikv.(*tikvTxn).Get\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/store/tikv/txn.go:122\ngithub.com/pingcap/tidb/structure.(*TxStructure).Get\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/structure/string.go:36\ngithub.com/pingcap/tidb/structure.(*TxStructure).GetInt64\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/structure/string.go:45\ngithub.com/pingcap/tidb/meta.(*Meta).GetBootstrapVersion\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/meta/meta.go:837\ngithub.com/pingcap/tidb/session.getStoreBootstrapVersion.func1\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/session/session.go:1797\ngithub.com/pingcap/tidb/kv.RunInNewTxn\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/kv/txn.go:47\ngithub.com/pingcap/tidb/session.getStoreBootstrapVersion\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/session/session.go:1794\ngithub.com/pingcap/tidb/session.BootstrapSession\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/session/session.go:1620\ngithub.com/pingcap/br/pkg/gluetidb.Glue.BootstrapSession\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/gluetidb/glue.go:29\ngithub.com/pingcap/br/pkg/conn.NewMgr\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/conn/conn.go:162\ngithub.com/pingcap/br/pkg/task.newMgr\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/task/common.go:230\ngithub.com/pingcap/br/pkg/task.RunBackup\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/task/backup.go:77\ngithub.com/pingcap/br/cmd.runBackupCommand\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/cmd/backup.go:18\ngithub.com/pingcap/br/cmd.newDbBackupCommand.func1\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/cmd/backup.go:71\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864\ngithub.com/pingcap/br.main\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/main.go:55\ngithub.com/pingcap/br.TestRunMain.func1\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/main_test.go:39\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357"] [stack="github.com/pingcap/tidb/session.getStoreBootstrapVersion\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/session/session.go:1802\ngithub.com/pingcap/tidb/session.BootstrapSession\n\t/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20200223044457-aedea3ec5e1e/session/session.go:1620\ngithub.com/pingcap/br/pkg/gluetidb.Glue.BootstrapSession\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/gluetidb/glue.go:29\ngithub.com/pingcap/br/pkg/conn.NewMgr\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/conn/conn.go:162\ngithub.com/pingcap/br/pkg/task.newMgr\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/task/common.go:230\ngithub.com/pingcap/br/pkg/task.RunBackup\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/pkg/task/backup.go:77\ngithub.com/pingcap/br/cmd.runBackupCommand\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/cmd/backup.go:18\ngithub.com/pingcap/br/cmd.newDbBackupCommand.func1\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/cmd/backup.go:71\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864\ngithub.com/pingcap/br.main\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/main.go:55\ngithub.com/pingcap/br.TestRunMain.func1\n\t/home/jenkins/agent/workspace/br_ghpr_unit_and_integration_test/go/src/github.com/pingcap/br/main_test.go:39"]
```

**Pull request merged**